archive/logic_puzzles/knights_and_knaves.lean

```diff
@@ -0,0 +1,86 @@
+/-
+Copyright (c) 2020 Dan Stanescu. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Dan Stanescu.
+-/
+
+import tactic
+
+/-!
+# Knights and knaves puzzles
+Two puzzles from "Knights and Knaves" by Raymond Smullyan.
+For an online description of these puzzles, see: http://mesosyn.com/mental1-6.html
+More information on R. Smullyan: https://en.wikipedia.org/wiki/Raymond_Smullyan
+
+There is an island where all inhabitants are either knights or knaves. 
+Knights always tell the truth. Knaves always lie.
+Joe and Bob are two inhabitants of this island.
+-/
+
+inductive islander
+| knight
+| knave
+
+notation `y` := islander.knight
+notation `n` := islander.knave
+
+structure Q := (Joe Bob : islander)
+
+/--
+In this first puzzle, only Joe makes a statement.
+He affirms that both protagonist (i.e. both Joe and Bob) are knaves.
+Question: what kind of islanders are they in fact?
+-/
+
+def Q1 := Q
+
+namespace Q1
+
+variables (q : Q1)
+
+-- Joe's statement; Joe is the first member of q
+def S1 := q.1 = n ∧ q.2 = n
+-- Stating that Joe can be a knight or a knave
+def H := (q.1 = y ∧ q.S1) ∨ (q.1 = n ∧ ¬ q.S1)
```

Comment:
Why write `q.S1` rather than just `S1`?

In fact, why even make the `def Q1 := Q`? It's not really the colloquial way to do things, and I'm not sure what it's gaining here. (I'm happy with an explanation, rather than a change, to be clear!)

Reply:
In my opinion, this better delineates the different puzzles. I don't think there's any drawback. It may just be a matter of taste.